pkg/services/ngalert/notifier/redis_peer.go

```diff
@@ -0,0 +1,568 @@
+		return 0
+	}
+	members, _ := scan.Val()
+	return len(members)
+}
+
+// If the cluster is healthy it should return 0, otherwise the number of
+// unhealthy nodes.
+func (p *redisPeer) GetHealthScore() int {
+	size := p.ClusterSize()
+	members := len(p.Members())
+	if size > members {
+		return size - members
+	}
+	return 0
+}
+
+// Members returns a list of active cluster Members.
+func (p *redisPeer) Members() []string {
+	p.membersMtx.Lock()
+	defer p.membersMtx.Unlock()
+	return p.members
+}
+
+func (p *redisPeer) WaitReady(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-p.readyc:
+		return nil
+	}
+}
+
+// Settle is mostly copied from uptream.
+// Ref: https://github.com/prometheus/alertmanager/blob/2888649b473970400c0bd375fdd563486dc80481/cluster/cluster.go#L674-L712
+func (p *redisPeer) Settle(ctx context.Context, interval time.Duration) {
+	const NumOkayRequired = 3
+	p.logger.Info("Waiting for gossip to settle...", "interval", interval)
+	start := time.Now()
+	nPeers := 0
+	nOkay := 0
+	totalPolls := 0
+	for {
+		select {
+		case <-ctx.Done():
+			elapsed := time.Since(start)
+			p.logger.Info("gossip not settled but continuing anyway", "polls", totalPolls, "elapsed", elapsed)
+			close(p.readyc)
+			return
+		case <-time.After(interval):
+		}
+		elapsed := time.Since(start)
+		n := len(p.Members())
+		if nOkay >= NumOkayRequired {
+			p.logger.Info("gossip settled; proceeding", "elapsed", elapsed)
+			break
+		}
+		if n == nPeers {
+			nOkay++
+			p.logger.Debug("gossip looks settled", "elapsed", elapsed)
+		} else {
+			nOkay = 0
+			p.logger.Info("gossip not settled", "polls", totalPolls, "before", nPeers, "now", n, "elapsed", elapsed)
+		}
+		nPeers = n
+		totalPolls++
+	}
+	p.requestFullState()
+	close(p.readyc)
+}
+
+func (p *redisPeer) AddState(key string, state cluster.State, _ prometheus.Registerer) cluster.ClusterChannel {
+	p.statesMtx.Lock()
+	defer p.statesMtx.Unlock()
+	p.states[key] = state
+	// As we also want to get the state from other nodes, we subscribe to the key.
+	sub := p.redis.Subscribe(context.Background(), p.withPrefix(key))
+	go p.receiveLoop(key, sub)
+	p.subs[key] = sub
+	return newRedisChannel(p, key, p.withPrefix(key), update)
+}
+
+func (p *redisPeer) receiveLoop(name string, channel *redis.PubSub) {
+	for {
+		select {
+		case <-p.shutdownc:
+			return
+		case data := <-channel.Channel():
+			p.mergePartialState([]byte(data.Payload))
+		default:
+			time.Sleep(waitForMsgIdle)
+		}
+	}
+}
+
+func (p *redisPeer) mergePartialState(buf []byte) {
+	p.messagesReceived.WithLabelValues(update).Inc()
+	p.messagesReceivedSize.WithLabelValues(update).Add(float64(len(buf)))
+
+	var part clusterpb.Part
+	if err := proto.Unmarshal([]byte(buf), &part); err != nil {
+		p.logger.Warn("error decoding the received broadcast message", "err", err)
+		return
+	}
+
+	p.statesMtx.RLock()
+	s, ok := p.states[part.Key]
+	p.statesMtx.RUnlock()
```

Comment:
Is `cluster.State` thread safe? I think we need to hold onto the lock here.

```suggestion
	p.statesMtx.RLock()
	s, ok := p.states[part.Key]
	defer p.statesMtx.RUnlock()
```